### PR TITLE
WIP: [Docs] Update extension type examples to not use UUID

### PR DESCRIPTION
**Pull request by khwilson**
Superseded by #44120

### CODE REVIEW

**Comment by ianmcook** [comment]
@khwilson I found one more docs page where there are a couple of mentions of UUID as an example of a user-defined extension type:

https://github.com/apache/arrow/blob/75ca5b3631144f58ea3edbe6b4933a686c0e0fd9/docs/source/format/Columnar.rst?plain=1#L1600-L1604

https://github.com/apache/arrow/blob/75ca5b3631144f58ea3edbe6b4933a686c0e0fd9/docs/source/format/Columnar.rst?plain=1#L1610

Could you replace these with your rational example? Thanks

File: docs/source/python/extending_types.rst [hunk 1]
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
...
+            return b""
 
         @classmethod
-        def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        def __arrow_ext_deserialize__(self, storage_type, serialized):
             # Sanity checks, not required but illustrate the method signature.
-            assert storage_type == pa.binary(16)
+            assert pt.is_struct(storage_type)
+            assert pt.is_int32(storage_type[0].type)
             assert serialized == b''
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:163]:
```suggestion
            assert serialized == b""
```

File: docs/source/python/extending_types.rst [hunk 2]
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
...
+                "my_package.rational",
+            )
+
+        def __arrow_ext_serialize__(self) -> bytes:
+            # No serialized metadata necessary
+            return b""
 
         @classmethod
-        def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        def __arrow_ext_deserialize__(self, storage_type, serialized):
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:159]:
```suggestion
        def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

File: docs/source/python/extending_types.rst [hunk 3]
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
...
             # Sanity checks, not required but illustrate the method signature.
-            assert storage_type == pa.binary(16)
+            assert pt.is_struct(storage_type)
+            assert pt.is_int32(storage_type[0].type)
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:162]: Might as well check denom too
```suggestion
            assert pt.is_int32(storage_type[0].type)
            assert pt.is_int32(storage_type[1].type)
```

File: docs/source/python/extending_types.rst [hunk 4]
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
...
 
-For example, we could define a custom UUID type for 128-bit numbers which can
-be represented as ``FixedSizeBinary`` type with 16 bytes::
+For example, we could define a custom rational type for fractions which can
+be represented as a pair of integers::
 
-    class UuidType(pa.ExtensionType):
+    import pyarrow as pa
+    import pyarrow.types as pt
+
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:139]: I think it's better to:
- omit these imports from the example
- just use `pa.types` in the code (instead of `pt`) here and in the other files

File: docs/source/python/extending_types.rst [hunk 5]
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
...
-            return UuidType()
+
+            # return an instance of this subclass given the serialized
+            # metadata
+            return RationalType()
+
 
 The special methods ``__arrow_ext_serialize__`` and ``__arrow_ext_deserialize__``
-define the serialization of an extension type instance. For non-parametric
-types such as the above, the serialization payload can be left empty.
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:158]: Why remove this?

File: docs/source/python/extending_types.rst [hunk 6]
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
...
+    'my_package.rational'
+    >>> rational_type.storage_type
+    StructType(struct<numer: int32, denom: int32>)
+
+    >>> storage_array = pa.array(
+    ... [
+    ...     {"numer": 10, "denom": 17},
+    ...     {"numer": 20, "denom": 13},
+    ... ],
+    ... type=rational_type.storage_type
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:186]: Please indent (like in the example in `types.pxi` below)

File: python/pyarrow/types.pxi [hunk 7]
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
...
+    ...                     ("denom", data_type),
+    ...                 ],
+    ...             ),
+    ...             # N.B. This name does _not_ reference `data_type` so deserialization
+    ...             # will work for _any_ integer `data_type` after registration
+    ...             "my_package.rational",
+    ...         )
+    ...     def __arrow_ext_serialize__(self) -> bytes:
+    ...         # No serialized metadata necessary
+    ...         return b""
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:1642]: This might be confusing to some readers. You're saying that even though this extension type is "parameterized" (in a sense), it does not need serialized metadata because it's self-describing (i.e. the parameters are known by looking at the storage type), correct? That is a subtle distinction that might confuse readers.

Could you take a stab at adding a sentence or two to describe this distinction in the section under the `Parameterized extension type` heading in `extending_types.rst` https://arrow.apache.org/docs/python/extending_types.html#parameterized-extension-type

File: docs/source/python/extending_types.rst (hunk 5 above)
Comment (ianmcook) [reply docs/source/python/extending_types.rst:158]: Oh, I see. I think this relates to what exactly parameterized / parametric means, correct?

- If the storage type requires one or more parameters to initially specify, does that qualify as "parameterized"?
- Or does "parameterized" mean that we need additional metadata (that is not encoded in the underlying storage type) in order to reconstruct the instance?

I think you are defining it to mean the latter, correct?

I think that makes sense. But it would be good to explicitly say this in the docs. See my other comment about that.

File: python/pyarrow/types.pxi [hunk 8]
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
...
-    >>> storage_array = pa.array([uuid.uuid4().bytes for _ in range(4)], pa.binary(16))
-    >>> uuid_type.wrap_array(storage_array)
+    >>> storage_array = pa.array(
+    ...     [
+    ...         {"numer": 10, "denom": 17},
+    ...         {"numer": 20, "denom": 13},
+    ...     ],
+    ...     type=rational_type.storage_type
+    ... )
+    >>> ratoinal_type.wrap_array(storage_array)
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:1673]: Small improvement to prevent readers from mistakenly thinking that `wrap_array` modifies `storage_array` in-place.
```suggestion
    >>> rational_array = ratoinal_type.wrap_array(storage_array)
    >>> rational_array
```

File: docs/source/python/extending_types.rst [hunk 9]
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
...
+    ... )
+    >>> arr = rational_type.wrap_array(storage_array)
+    >>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:189]:
```suggestion
    >>> arr = rational_type.wrap_array(storage_array)
    # or equivalently
    >>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
```

File: python/pyarrow/types.pxi [hunk 10]
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
...
     <pyarrow.lib.ExtensionArray object at ...>
-    [
+    -- is_valid: all not null
       ...
-    ]
 
     Or do the same with creating an ExtensionArray:
 
-    >>> pa.ExtensionArray.from_storage(uuid_type, storage_array)
+    >>> pa.ExtensionArray.from_storage(rational_type, storage_array)
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:1680]:
```suggestion
    >>> rational_array = pa.ExtensionArray.from_storage(rational_type, storage_array)
    >>> rational_array
```

File: python/pyarrow/types.pxi [hunk 11]
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
...
+    ...             ),
+    ...             # N.B. This name does _not_ reference `data_type` so deserialization
+    ...             # will work for _any_ integer `data_type` after registration
+    ...             "my_package.rational",
+    ...         )
+    ...     def __arrow_ext_serialize__(self) -> bytes:
+    ...         # No serialized metadata necessary
+    ...         return b""
+    ...     @classmethod
+    ...     def __arrow_ext_deserialize__(self, storage_type, serialized):
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:1644]:
```suggestion
    ...     def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

File: python/pyarrow/types.pxi [hunk 12]
```diff
@@ -2089,30 +2120,41 @@ def unregister_extension_type(type_name):
...
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:2146]:
```suggestion
    ...     def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

File: python/pyarrow/types.pxi [hunk 13]
```diff
@@ -2039,30 +2059,41 @@ def register_extension_type(ext_type):
...
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:2085]:
```suggestion
    ...     def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

**Comment by ianmcook** [comment]
@khwilson thank you; this looks pretty good! I left a bunch of comments. They should be mostly be easy to address (except the bit about what "parameterized" means which might require some thinking about how best to explain it).

File: docs/source/python/extending_types.rst (hunk 1 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:163]: A question: There are several other places in this file where single quotes were used, e.g., https://github.com/apache/arrow/blob/fde32158fa5f9a5f943da5aacd2bd8056b7a5bb1/docs/source/python/extending_types.rst?plain=1#L267

Would it be helpful to add an issue to standardize to double quotes?

File: docs/source/python/extending_types.rst (hunk 4 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:139]: Done

File: docs/source/python/extending_types.rst (hunk 6 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:187]: Done!

File: python/pyarrow/types.pxi (hunk 8 above)
Comment (khwilson) [reply python/pyarrow/types.pxi:1673]: Done.

File: docs/source/python/extending_types.rst (hunk 9 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:190]: This is the same as the original documentation. I'm guessing the previous writer didn't insert `# or equivalently` as as it causes `doctest` to fail. I've changed this to the following:

```python
"""
>>> arr = rational_type.wrap_array(storage_array)
>>> # or equivalently
>>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
"""
```

But the alternative (which I think is harder to read) is:

```python
"""
>>> arr = rational_type.wrap_array(storage_array)

or equivalently

>>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
"""
```

File: python/pyarrow/types.pxi (hunk 11 above)
Comment (khwilson) [reply python/pyarrow/types.pxi:1644]: fwiw, the `self` is actually in the function signature for `__arrow_ext_deserialize__` on L1753. I've modified that signature as well :-)

File: python/pyarrow/types.pxi (hunk 12 above)
Comment (khwilson) [reply python/pyarrow/types.pxi:2146]: Done

File: python/pyarrow/types.pxi (hunk 13 above)
Comment (khwilson) [reply python/pyarrow/types.pxi:2085]: Done

File: python/pyarrow/types.pxi (hunk 10 above)
Comment (khwilson) [reply python/pyarrow/types.pxi:1680]: Done

File: docs/source/python/extending_types.rst (hunk 1 above)
Comment (ianmcook) [reply docs/source/python/extending_types.rst:164]: > Would it be helpful to add an issue to standardize to double quotes?

Sure, or if you just want to fix them in this PR, that's fine too. I don't think we're too fussy about consistency of single/double quotes in examples, but it's best to at least be consistent within one example.

File: docs/source/python/extending_types.rst (hunk 9 above)
Comment (ianmcook) [reply docs/source/python/extending_types.rst:190]: Thanks. That looks good to me.

**Comment by khwilson** [comment]
Thanks for the review, @ianmcook! I've addressed all the minor comments and will take another pass at the "parameterized" definition on the cross continental flight I'm about to get on.

File: docs/source/python/extending_types.rst (hunk 1 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:161]: Fair! I'll go ahead and change them all while I'm here.

**Comment by khwilson** [comment]
Apologies for my delay, @ianmcook . Ended up just taking a nap on that flight. :-)

I've taken a stab at disambiguating "parameterization." To do so, I decided to explicitly call the metadata captured by the `ARROW:extension:metadata` field "parameters." If this is outside the bounds of a PR and should be discussed more broadly, please let me know.

The only remaining unaddressed comment from your review about the UUID type in Columnar.rst. However, I actually don't read that one as needing to be replaced _unless_ the intention of _canonical_ extension types is that all Arrow implementations _must_ or _should_ support them. Reading the page on [canonical extensions](https://arrow.apache.org/docs/format/CanonicalExtensions.html) it is unclear to me if arrow implementations _must_ , _should_ , or _may_ support them. If it is _may_, then I think the text of Columnar is actually quite good, though I might modify it as follows:

> For example, Arrow specifies a canonical extension type that represents a UUID as a ``FixedSizeBinary(16)``. Arrow implementations are not required to support canonical extensions, so an implementation that does not support this UUID type will simply interpret it as a ``FixedSizeBinary(16)`` and pass along the ``custom_metadata`` in subsequent Arrow protocol messages.

**Comment by ianmcook** [comment]
Thanks @khwilson! I will re-review later this week.

File: docs/source/python/extending_types.rst [hunk 14]
```diff
@@ -116,73 +116,103 @@ a :class:`~pyarrow.Array` or a :class:`~pyarrow.ChunkedArray`.
...
-:class:`ExtensionType` and giving the derived class its own extension name
-and serialization mechanism. The extension name and serialized metadata
-can potentially be recognized by other (non-Python) Arrow implementations
+Arrow affords a notion of extension types which allow users to annotate data
+types with additional semantics. This allows downstream consumers both to
+specify custom serialization and deserialization routines (for example,
+to :ref:`Python scalars <custom-scalar-conversion>` and
+:ref:`pandas <conversion-to-pandas>`) and to more easily interpret data.
+
+In the Arrow :doc:`metadata specification</format/Columnar.html#format-metadata-extension-types>`,
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:125]: Please use the existing ref. 
```suggestion
In the Arrow :ref:`format_metadata_extension_types`,
```

This points here:
https://github.com/apache/arrow/blob/ab0a40ee34217070f14027776682074c55d0b507/docs/source/format/Columnar.rst?plain=1#L1575-L1578

File: docs/source/python/extending_types.rst [hunk 15]
```diff
@@ -116,73 +116,103 @@ a :class:`~pyarrow.Array` or a :class:`~pyarrow.ChunkedArray`.
...
-message).
-See the :ref:`format_metadata_extension_types` section of the metadata
-specification for more details.
-
-Pyarrow allows you to define such extension types from Python by subclassing
-:class:`ExtensionType` and giving the derived class its own extension name
-and serialization mechanism. The extension name and serialized metadata
-can potentially be recognized by other (non-Python) Arrow implementations
+Arrow affords a notion of extension types which allow users to annotate data
+types with additional semantics. This allows downstream consumers both to
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:120]: I'm -1 on the "downstream consumers" phrasing. How about just "developers"?

File: docs/source/python/extending_types.rst [hunk 16]
```diff
@@ -197,19 +227,44 @@ and then reading it back yields the proper type::
...
     ...    result = reader.read_all()
-    >>> result.column('ext').type
-    UuidType(FixedSizeBinaryType(fixed_size_binary[16]))
+    >>> result.column("ext").type
+    RationalType(StructType(struct<numer: int32, denom: int32>))
+
+Further, note that while we registered the concrete type
+``RationalType(pa.int32())``, ``RationalType(integer_type)`` has the same
+extension name (``"my_package.rational"``) for all integer types. As such,
+the above code will also allow users to (de)serialize these data types::
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:236]: Edit for readability
```suggestion
``RationalType(pa.int32())``, you could use the same extension name
``"my_package.rational"`` to register ``RationalType(integer_type)`` for
any Arrow integer type. As such, the above code will also allow users to
(de)serialize these data types::
```

File: docs/source/python/extending_types.rst [hunk 17]
```diff
@@ -116,73 +116,103 @@ a :class:`~pyarrow.Array` or a :class:`~pyarrow.ChunkedArray`.
...
-    class UuidType(pa.ExtensionType):
-
-        def __init__(self):
-            super().__init__(pa.binary(16), "my_package.uuid")
-
-        def __arrow_ext_serialize__(self):
-            # Since we don't have a parameterized type, we don't need extra
-            # metadata to be deserialized
-            return b''
+Pyarrow allows you to define extension types from Python by subclassing
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:137]:
```suggestion
PyArrow allows you to define extension types from Python by subclassing
```

File: docs/source/python/extending_types.rst [hunk 18]
```diff
@@ -225,22 +280,22 @@ of the given frequency since 1970.
...
-            super().__init__(pa.int64(), 'my_package.period')
+            super().__init__(pa.int64(), "my_package.period")
 
         @property
         def freq(self):
             return self._freq
 
         def __arrow_ext_serialize__(self):
-            return "freq={}".format(self.freq).encode()
+            return "freq={self.freq}".encode()
```
Comment (ianmcook) [inline docs/source/python/extending_types.rst:290]: Please revert the change to this one line. It was correct before but now it is not.

File: python/pyarrow/types.pxi [hunk 19]
```diff
@@ -1618,59 +1618,78 @@ cdef class ExtensionType(BaseExtensionType):
...
-    ]
 
     Unregister the extension type:
 
-    >>> pa.unregister_extension_type("my_package.uuid")
+    >>> pa.unregister_extension_type("my_package.rational")
+
+    Note that even though we registered the concrete type
+    ``RationalType(pa.int64())``, pyarrow will be able to deserialize
+    ``RationalType(integer_type)`` for any ``integer_type`` as the deserializer
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:1690]:
```suggestion
    ``RationalType(pa.int64())``, PyArrow will be able to deserialize
    ``RationalType(integer_type)`` for any ``integer_type``, as the deserializer
```

**Comment by ianmcook** [comment]
@khwilson I made some suggested changes. Once you've incorporated these I'll do a final pass and then get some help to figure out why the CI is still failing.

**Comment by ianmcook** [comment]
> If this is outside the bounds of a PR and should be discussed more broadly, please let me know.
No, I think it's good. Thank you.

**Comment by ianmcook** [comment]
> The only remaining unaddressed comment from your review about the UUID type in Columnar.rst. However, I actually don't read that one as needing to be replaced _unless_ the intention of _canonical_ extension types is that all Arrow implementations _must_ or _should_ support them. Reading the page on [canonical extensions](https://arrow.apache.org/docs/format/CanonicalExtensions.html) it is unclear to me if arrow implementations _must_ , _should_ , or _may_ support them. If it is _may_, then I think the text of Columnar is actually quite good, though I might modify it as follows:
> 
> > For example, Arrow specifies a canonical extension type that represents a UUID as a `FixedSizeBinary(16)`. Arrow implementations are not required to support canonical extensions, so an implementation that does not support this UUID type will simply interpret it as a `FixedSizeBinary(16)` and pass along the `custom_metadata` in subsequent Arrow protocol messages.

I like your suggested modification. I think it's fine that we refer to a canonical extension type there, but I think it's worth saying explicitly that it's a canonical extension type, which is exactly what your modified text does. 👍

File: docs/source/python/extending_types.rst (hunk 15 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:120]: Fair. I had in mind that end users might also come to rely on metadata (e.g., you might use this extension type to create an SPSS-like survey data format), but this seems like a reasonably niche use case.

File: docs/source/python/extending_types.rst (hunk 14 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:125]: Thanks! Missed that reference.

File: docs/source/python/extending_types.rst (hunk 16 above)
Comment (khwilson) [reply docs/source/python/extending_types.rst:236]: Adopted with a small change.

**Comment by khwilson** [comment]
Thanks, @ianmcook . Addressed your comments and I think I figured out at least one reason the doctests were failing and have fixed that.

**Comment by ianmcook** [comment]
@github-actions crossbow submit preview-docs

**Comment by ianmcook** [comment]
@github-actions crossbow submit preview-docs

**Comment by ianmcook** [comment]
@khwilson friendly nudge to please sync your fork with upstream main to see if that helps resolve the errors. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

**Comment by khwilson** [comment]
Hi @ianmcook i did do the sync and also setup archery locally to see if I could do some debugging separate from the long GA flow. I’ve resolved one issue (the missing “unregister” call) but another issue popped up: the doctests.ELLIPSIS matcher started failing on *other* tests.

Perhaps I’m missing some local environment variable setup?

File: python/pyarrow/types.pxi [hunk 20]
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
...
+    ...         {"numer": 20, "denom": 13},
+    ...     ],
+    ...     type=rational_type.storage_type
+    ... )
+    >>> rational_array = rational_type.wrap_array(storage_array)
+    >>> rational_array
     <pyarrow.lib.ExtensionArray object at ...>
-    [
+    -- is_valid: all not null
       ...
```
Comment (ianmcook) [inline python/pyarrow/types.pxi:1675]: It looks like one of the test failures is happening because this shows the ellipsis instead of the full output: https://github.com/apache/arrow/actions/runs/10762310741/job/29849821115?pr=43849#step:6:9145

**Comment by ianmcook** [comment]
> Hi @ianmcook i did do the sync and also setup archery locally to see if I could do some debugging separate from the long GA flow. I’ve resolved one issue (the missing “unregister” call) but another issue popped up: the doctests.ELLIPSIS matcher started failing on _other_ tests.
> 
> Perhaps I’m missing some local environment variable setup?

Can you try adding doctest directives like `  # doctest: +ELLIPSIS` to fix that issue?
 https://docs.python.org/3/library/doctest.html#directives

**Comment by khwilson** [comment]
Hi all. My deepest apologies. I got myself into a bit of trouble syncing with master and accidentally rebased instead of merging. I will close this issue for now to avoid spamming everyone as there is no way I know to unsubscribe you automatically.

**Comment by ianmcook** [comment]
No worries, it happens :) Tag me in the new PR and I will review
